File: frontend/src/lib/api/stats.ts
import request from './request';

export interface StatusBreakdown {
	not_started: number;
	pending: number;
	processing: number;
	done: number;
	failed: number;
}

export interface RecentDocument {
	id: number;
	title: string | null;
	original_file_name: string;
	file_size: number;
	enrich_status: string;
	created_at: string;
}

export interface DayCount {
	date: string;
	count: number;
}

export interface FormatBucket {
	content_type: string;
	count: number;
}

export interface TopDocument {
	id: number;
	title: string | null;
	original_file_name: string;
	view_count: number;
	like_count: number;
}

export interface Contributor {
	user_id: number;
	username: string;
	nickname: string | null;
	avatar_url: string | null;
	doc_count: number;
}

export interface DashboardStatsResponse {
	total_documents: number;
	total_storage: number;
	total_views: number;
	total_likes: number;
	status_breakdown: StatusBreakdown;
	recent_documents: RecentDocument[];
	uploads_by_day: DayCount[];
	views_by_day: DayCount[];
	likes_by_day: DayCount[];
	format_distribution: FormatBucket[];
	top_viewed: TopDocument[];
}

export interface LabDashboardStatsResponse {
	total_documents: number;
	total_storage: number;
	total_views: number;
	total_likes: number;
	member_count: number;
	status_breakdown: StatusBreakdown;
	recent_documents: RecentDocument[];
	uploads_by_day: DayCount[];
	views_by_day: DayCount[];
	likes_by_day: DayCount[];
	format_distribution: FormatBucket[];
	top_contributors: Contributor[];
}

const statsApi = {
	getDashboardStats(): Promise<DashboardStatsResponse> {
		return request.get<DashboardStatsResponse>(
			'/stats/mine/dashboard'
		) as unknown as Promise<DashboardStatsResponse>;
	},
	getLabDashboardStats(labID: number): Promise<LabDashboardStatsResponse> {
		return request.get<LabDashboardStatsResponse>(
			`/stats/labs/${labID}/dashboard`
		) as unknown as Promise<LabDashboardStatsResponse>;
	}
};

export default statsApi;

File: svc-gateway/internal/dto/stats.go
package dto

import "time"

type StatusBreakdown struct {
	NotStarted int64 `json:"not_started"`
	Pending    int64 `json:"pending"`
	Processing int64 `json:"processing"`
	Done       int64 `json:"done"`
	Failed     int64 `json:"failed"`
}

type RecentDocument struct {
	ID               uint      `json:"id"`
	Title            *string   `json:"title"`
	OriginalFileName string    `json:"original_file_name"`
	FileSize         int64     `json:"file_size"`
	EnrichStatus     string    `json:"enrich_status"`
	CreatedAt        time.Time `json:"created_at"`
}

// DayCount is one bucket of a daily time series. Date is YYYY-MM-DD in UTC.
type DayCount struct {
	Date  string `json:"date"`
	Count int64  `json:"count"`
}

// FormatBucket is one slice of the format-distribution donut.
type FormatBucket struct {
	ContentType string `json:"content_type"`
	Count       int64  `json:"count"`
}

// TopDocument is a row in the "top viewed documents" list shown on the user dashboard.
type TopDocument struct {
	ID               uint    `json:"id"`
	Title            *string `json:"title"`
	OriginalFileName string  `json:"original_file_name"`
	ViewCount        int64   `json:"view_count"`
	LikeCount        int64   `json:"like_count"`
}

// Contributor is a row in the "top contributors" list shown on the lab dashboard.
type Contributor struct {
	UserID    uint    `json:"user_id"`
	Username  string  `json:"username"`
	Nickname  *string `json:"nickname"`
	AvatarURL *string `json:"avatar_url"`
	DocCount  int64   `json:"doc_count"`
}

type MyDashboardStatsResponse struct {
	TotalDocuments     int64            `json:"total_documents"`
	TotalStorage       int64            `json:"total_storage"`
	TotalViews         int64            `json:"total_views"`
	TotalLikes         int64            `json:"total_likes"`
	StatusBreakdown    StatusBreakdown  `json:"status_breakdown"`
	RecentDocuments    []RecentDocument `json:"recent_documents"`
	UploadsByDay       []DayCount       `json:"uploads_by_day"`
	ViewsByDay         []DayCount       `json:"views_by_day"`
	LikesByDay         []DayCount       `json:"likes_by_day"`
	FormatDistribution []FormatBucket   `json:"format_distribution"`
	TopViewed          []TopDocument    `json:"top_viewed"`
}

type LabDashboardStatsResponse struct {
	TotalDocuments     int64            `json:"total_documents"`
	TotalStorage       int64            `json:"total_storage"`
	TotalViews         int64            `json:"total_views"`
	TotalLikes         int64            `json:"total_likes"`
	MemberCount        int64            `json:"member_count"`
	StatusBreakdown    StatusBreakdown  `json:"status_breakdown"`
	RecentDocuments    []RecentDocument `json:"recent_documents"`
	UploadsByDay       []DayCount       `json:"uploads_by_day"`
	ViewsByDay         []DayCount       `json:"views_by_day"`
	LikesByDay         []DayCount       `json:"likes_by_day"`
	FormatDistribution []FormatBucket   `json:"format_distribution"`
	TopContributors    []Contributor    `json:"top_contributors"`
}
